File: RottingOranges.cpp
// Time Complexity: O(m*n)
// Space Compleixty: O(m+n)

class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        deque<vector<int>> q; 
        int freshOranges = 0; 

        for(int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[0].size(); j++) {
                if(grid[i][j] == 2) {
                    q.push_back({i,j});
                }
            }
        }

        int elapsedTime = rottingOranges(q, grid);

        for(int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[0].size(); j++) {
                if(grid[i][j] == 1) freshOranges++; 
            }
        }


        return freshOranges == 0 ? elapsedTime : -1;         
    }

    int rottingOranges(deque<vector<int>>& q, vector<vector<int>>& grid) {
        int time = 0; 

        int dir[4][2] = { {-1,0}, {1,0}, {0,-1}, {0,1} };

        while(!q.empty()) {
            int n = q.size(); 
            bool flag = false;

            for(int i = 0; i < n; i++) {
                int row = q.front()[0];
                int col = q.front()[1];
                q.pop_front(); 

                for(int di = 0; di < 4; di++) {
                    int new_row = row + dir[di][0];
                    int new_col = col + dir[di][1];

                    if(new_row >= 0 && new_col >= 0 && new_row < grid.size() && new_col < grid[0].size()) {
                        if(grid[new_row][new_col] == 1) {
                            grid[new_row][new_col] = 2; 
                            q.push_back({new_row, new_col});
                            flag = true; 
                        }
                    }
                }     
            }

            if(flag == true) time++;  

        }

        return time; 
    }
};

File: rottingoranges.java

// Time Complexity : O(mn)
// Space Complexity : O(mn)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this : no

public class rottingoranges {
        public static int
                EMPTY = 0,
                FRESH = 1,
                ROTTEN = 2;

        public int directions[][] = {
                {-1, 0},
                {0, 1},
                {1, 0},
                {0 , -1}
        };
        public int orangesRotting(int[][] grid) {

            int m = grid.length;
            int n = grid[0].length;
            int timer = 0;
            int frshOranges = 0;

            Queue <Pair> q = new LinkedList<>();

            for (int i = 0 ; i < m ; ++i)  {
                for (int j = 0 ; j < n ; ++j){
                    if(grid[i][j] == ROTTEN){
                        q.add(new Pair(i,j));
                    }
                    else if(grid[i][j] == FRESH){
                        frshOranges++;
                    }
                }
            }


            while(!q.isEmpty()){

                Queue <Pair> qt = new LinkedList<>();

                while(!q.isEmpty()){
                    Pair p = q.poll();

                    for(int [] dir : directions){
                        int x = p.x + dir[0];
                        int y = p.y + dir[1];

                        if(x >= 0 && x < m && y >= 0 && y < n && grid[x][y] == FRESH){
                            grid[x][y] = ROTTEN;
                            frshOranges--;
                            qt.add(new Pair(x,y));
                        }
                    }
                }
                q = qt;
                if(!qt.isEmpty()){
                    timer++;
                }
            }
            return frshOranges == 0 ? timer : -1;
        }
    }

    class Pair{
        int x;
        int y;
        Pair(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
